SleepEngine/Engine/Components/HierarchyNode.cpp

```diff
@@ -7,57 +7,75 @@ BEGIN_NAMESPACE_SLEEP
 
 HierarchyNode::HierarchyNode()
     : Base(Component::getComponentTypeID<HierarchyNode>())
+    , m_parent(nullptr)
 {}
 
 Object* HierarchyNode::addChild(std::unique_ptr<Object>&& child)
 {
-    auto* childNode = child->getComponent<HierarchyNode>();
+    auto* const childHandle = child.get();
 
-    if(!childNode)
-    {
-        childNode = child->addComponent<HierarchyNode>();
-    }
-
-    if (childNode->getParent())
-    {
-        childNode->detachFromParent();
-    }
-    childNode->setParent(m_object);
+    setUpChild(*childHandle);
 
-    auto* childHandle = child.get();
     m_children.emplace_back(std::move(child));
     return childHandle;
 }
 
 void HierarchyNode::removeChild(Object* child)
 {
-    auto isChildToDelete = [child] (std::unique_ptr<Object> const& element)
+    auto const isChildToDelete = [child] (std::unique_ptr<Object> const& element)
     {
         return element.get() == child;
     };
-    LOG_AND_ASSERT_ERROR(containsIf(m_children, isChildToDelete), "not my child!");
-    removeIf(m_children, isChildToDelete);
+
+    auto const childToDelete = findIf(m_children, isChildToDelete);
+
+    if (childToDelete != m_children.cend())
+    {
+        LOG_AND_FAIL("'removeChild' called for object that is not out child");
+        return;
+    }
+
+    m_children.erase(childToDelete);
 }
 
 void HierarchyNode::detachFromParent()
 {
-    if (!m_object)
+    if (!m_parent)
     {
         LOG_AND_FAIL("parent is null");
         return;
     }
 
-    auto* parentNode = m_object->getComponent<HierarchyNode>();
-    parentNode->removeChild(m_object);
-    m_object = nullptr;
+    auto* const parentNode = m_parent->getComponent<HierarchyNode>();
+    parentNode->removeChild(m_parent);
+    m_parent = nullptr;

```

Comment:
Let's init the trivial members inline& its what we both were doing so far from what i could see

Reply:
But this is not an initialization

Reply:
oh, meant to comment other line 
![image](https://user-images.githubusercontent.com/23706483/56870051-d0b4aa00-6a11-11e9-83bc-e4ef458eb1bb.png)

Reply:
I just don't like any initialization in header file if we have .cpp.
One thing to do it if it is a simple struct. But in class with interface, it looks like unnecessary information in the header.
In my opinion, it's pretty ok for Transform2DData, for example, but not for this class.

Reply:
maybe, but it is easier to track that way. If you have more than 1 constructor
plus, you don't need default constructor just to init the variables

Reply:
Plus, the members are implementation detail anyway.

Reply:
Sure. Fixed